examples/mnist_dataset_api.py

```diff
@@ -0,0 +1,116 @@
+'''MNIST classification with TensorFlow's Dataset API.
+
+Introduced in TensorFlow 1.3, the Dataset API is now the
+standard method for loading data into TensorFlow models.
+A Dataset is a sequence of elements, which are themselves
+composed of tf.Tensor components. For more details, see:
+https://www.tensorflow.org/programmers_guide/datasets
+
+To use this with Keras, we make a dataset out of elements
+of the form (input batch, output batch). From there, we
+create a one-shot iterator and a graph node corresponding
+to its get_next() method. Its components are then provided
+to the network's Input layer and the Model.compile() method,
+respectively.
+
+Note that from TensorFlow 1.4, tf.contrib.data is deprecated
+and tf.data is preferred. See the release notes for details.
+
+This example is intended to closely follow the
+mnist_tfrecord.py example. No effort is made for Python 2
+compatibility.
+'''
+import numpy as np
+import os
+import tempfile
+
+import keras
+from keras import backend as K
+from keras import layers
+from keras.datasets import mnist
+
+import tensorflow as tf
+from tensorflow.contrib.data import Dataset
+
+
+if K.backend() != 'tensorflow':
+    raise RuntimeError('This example can only run with the TensorFlow backend,'
+                       ' because it requires the Datset API, which is not'
+                       ' supported on other platforms.')
+
+
+def cnn_layers(x_train_input):
+    x = layers.Conv2D(32, (3, 3),
+                      activation='relu', padding='valid')(x_train_input)
+    x = layers.MaxPooling2D(pool_size=(2, 2))(x)
+    x = layers.Conv2D(64, (3, 3), activation='relu')(x)
+    x = layers.MaxPooling2D(pool_size=(2, 2))(x)
+    x = layers.Flatten()(x)
+    x = layers.Dense(512, activation='relu')(x)
+    x = layers.Dropout(0.5)(x)
+    x_train_out = layers.Dense(classes,
+                               activation='softmax',
+                               name='x_train_out')(x)
+    return x_train_out
+
+
+batch_size = 128
+buffer_size = 10000
+steps_per_epoch = np.ceil(60_000 / 128).astype('int')  # = 469
+epochs = 5
+classes = 10
+
+(x_train, y_train), (x_test, y_test) = mnist.load_data()
+x_train = x_train.astype(np.float32) / 255
+x_train = np.expand_dims(x_train, -1)
+y_train = tf.one_hot(y_train, classes)
```

Comment:
You don't use the `_train` arrays so you can `_` them altogether.

Also: please put this block when you need it, i.e. at test time (see comment below)

Reply:
Sorry, I find this comment a bit confusing. I use the `_train` arrays when I create the `dataset` object a few lines below. If I `_` them out, how will I get the training data to the model?

Would you prefer I have two separate calls to `mnist.load_data()`: one for the training data here and the other for the test data in the testing phase?